### PR TITLE
[dv/top] Add usb clock calibration test

### PR DESCRIPTION
**Pull request by tjaychen**
- addresses #14137 
- fake the usb portions since usb chip level test is not ready
- measurement the clocks pre/post calibration to ensure it is
  as expected
- tweak the ast logic slightly such that we can introduce a
  large "drift" using run time options

Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (tjaychen) [inline hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:50]: @sha-ron 
@Jacob-Levy 
@arnonsha 

hey everyone, it turns out the usb reference calibration logic doesn't really do anything in the open source AST (ie, the pulse is unused, only the valid is used). I think if possible, it would be a good idea to enhance this a bit so the open source side has more coverage. 

I hacked in an option for now that would introduce a large "drift", this makes the usb clock calibration testing more straightforward. Let me know how you guys prefer to handle this, this is just an example.

**Comment by tjaychen** [comment]
@a-will let me know if you think i've picked a reasonable force point.  I wanted to go further deeper, but then I ran into a bunch of packet completion and valid logic.

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:50]: It can be done. A special mecanism that fix the drift using the clock count between pulses need to be devise.

Comment (tjaychen) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:50]: yeah i'll leave it up to you guys to decide what's okay to expose in open source.
In the mean time, can you have a look at the "large drift" option i've added and see if it's okay on your end?

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:50]: Will review it soon...

**Comment by a-will** [comment]
> @a-will let me know if you think i've picked a reasonable force point. I wanted to go further deeper, but then I ran into a bunch of packet completion and valid logic.

I think what you have is reasonable, though this might be a little closer:

1. Ensure sense is high (pinmux config, set to tie-high) and usbdev is configured to connect (enable via CSR)
2. Suppress usb_fs_rx.rx_idle_det_o (to avoid suspending, force to 0)
3. Change usbdev_linkstate.link_state_q to LinkActiveNoSOF or LinkActive
4. Force usb_fs_nb_pe.sof_valid_o and change usb_fs_rx.frame_num_q when you want to pulse SOF

**Comment by tjaychen** [comment]
thanks @a-will ! i'll give that a try.

**Comment by tjaychen** [comment]
i'm going to tweak the test following @a-will suggestions. 
@Jacob-Levy can you have a look at the ast related changes?

**Comment by Jacob-Levy** [comment]
From the last point of view, the 1ms reference window is used to count the USB clock and fix the drift. It usually takes 3 to 6 valid pulses to fix the drift. 
Will review the changes of the usb_osc.sv to see if it does the work. BTW, the drift can be positive and negative.

**Comment by tjaychen** [comment]
> From the last point of view, the 1ms reference window is used to count the USB clock and fix the drift. It usually takes 3 to 6 valid pulses to fix the drift. Will review the changes of the usb_osc.sv to see if it does the work. BTW, the drift can be positive and negative.

yep understood. so i'm not sure if you guys want to actually model the fact that it takes a few pulses to correct. The pulse info isn't used right now, so after the first pulse the clock just becomes highly accurate. 

The only reason to model it, in my mind, is so that when this test is run with proprietary AST there arent any timing discrepancies.

**Comment by Jacob-Levy** [comment]
Don't limit the test to one pulse, I would like to run the test with the real AST too.

**Comment by tjaychen** [comment]
> Don't limit the test to one pulse, I would like to run the test with the real AST too.

yes that's right. The test i think does 5 right now, but my point is more I have no idea how many it should be.
If you expose a parameter via `ast.hjson`, something like `MIN_PULSES`, then that can be translated into a C header, and I can use that directly in the test.

**Comment by Jacob-Levy** [comment]
The number of pulse needed depend on the drift size direction and process. 3-6 pulses is typical convergence. The USB clock spec say 30 or 50 ms to fix the drift...
Let me review your code and see if I can come up with something that can work for the real AST too.

**Comment by tjaychen** [comment]
> The number of pulse needed depend on the drift size direction and process. 3-6 pulses is typical convergence. The USB clock spec say 30 or 50 ms to fix the drift... Let me review your code and see if I can come up with something that can work for the real AST too.

sounds good, thanks Jacob.

**Comment by tjaychen** [comment]
@a-will please have a look. I changed the forcing points. I tried to avoid directly forcing the state.
In the future I can probably the se0 forcing slightly as well.  But this seems to do the trick for now.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_usb_ast_clk_calib_vseq.sv
Comment (a-will) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_usb_ast_clk_calib_vseq.sv:27]: Heh, bonus points for taking us through the state machine with the bus reset :joy:

**Review by a-will — COMMENTED** [review]
The forcing points look okay to me.

**Review by engdoreis — COMMENTED** [review]
Thanks @tjaychen for the PR. I left some comments.

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (tjaychen) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:50]: @Jacob-Levy can you confirm if this change is okay to add? 
Of course you can enhance it how you see fit later, but want to make sure it doesn't cause a conflict.

File: sw/device/tests/sim_dv/ast_usb_clk_calib.c
Comment (tjaychen) [reply sw/device/tests/sim_dv/ast_usb_clk_calib.c:18]: o this was honestly just copy / pasted over... let me remove it.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_usb_ast_clk_calib_vseq.sv
Comment (tjaychen) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_usb_ast_clk_calib_vseq.sv:73]: fixed.

File: sw/device/tests/sim_dv/ast_usb_clk_calib.c
Comment (tjaychen) [reply sw/device/tests/sim_dv/ast_usb_clk_calib.c]: i've removed this entire comment altogether.

File: sw/device/tests/sim_dv/ast_usb_clk_calib.c
Comment (tjaychen) [reply sw/device/tests/sim_dv/ast_usb_clk_calib.c]: fixed.

File: sw/device/tests/sim_dv/ast_usb_clk_calib.c
Comment (tjaychen) [reply sw/device/tests/sim_dv/ast_usb_clk_calib.c]: that sounds good to me, but i was thinking i'd leave it to other sw folks to decide, and i can move it over if everyone thinks that's the best approach.

File: sw/device/tests/sim_dv/ast_usb_clk_calib.c
Comment (tjaychen) [reply sw/device/tests/sim_dv/ast_usb_clk_calib.c]: yep, i've made them all consistent.

**Comment by tjaychen** [comment]
please have a look and approve if it looks good.

File: hw/top_earlgrey/data/chip_testplan.hjson
Comment (tjaychen) [reply hw/top_earlgrey/data/chip_testplan.hjson]: i've converted to bullet points.

**Comment by Jacob-Levy** [comment]
@tjaychen I reviewed the changes you made and I need to make some changes in order to emulate the functiion of the beacon.
The usb_ref_val_i only qualifies the usb_ref_pulse_i. I will update the usb_osc.sv model to reflect the functionality of beacon.

Screenshot 2022-08-22 225907

**Comment by tjaychen** [comment]
> @tjaychen I reviewed the changes you made and I need to make some changes in order to emulate the functiion of the beacon. The usb_ref_val_i only qualifies the usb_ref_pulse_i. I will update the usb_osc.sv model to reflect the functionality of beacon.
> 
> Screenshot 2022-08-22 225907

thanks @Jacob-Levy . do you want me to hold off on merging this until you have a chance to introduce your change?

**Comment by Jacob-Levy** [comment]
I am not clear how the test verify the beacon function, but I need more time to describe the model and make sure it will represent the real AST too.

**Comment by tjaychen** [comment]
> I am not clear how the test verify the beacon function, but I need more time to describe the model and make sure it will represent the real AST too.

so what the test currently does is that it "fakes" the usb sof function, and sends periodic 1ms time-stamps to the ast. Since right not, the ast instantly corrects the clocks, it just waits for that to happen, and then does a clock frequency measurement on the usb clock.  So the frequency goes from "way off" to "correct".

**Review by engdoreis — APPROVED** [review]
Thanks @tjaychen for the fixes. The SW code LGTM.

**Comment by Jacob-Levy** [comment]
@tjaychen In order for the `usb_ref_pulse_i & usb_ref_val_i` affect to be checked, the usb_clk.sv & the usb_osc.sv need to be changed. At the moment,  the `usb_ref_pulse_i` that goes into the usb_clk.sv is not used. I used only the `usb_ref_val_i` that 
goes to the usb_osc.sv and force calibration when set to '1'. To make the calibration depend on the  `usb_ref_pulse_i too, I need to add it to the usb_osc.sv and force the calibration only when at least 'N' pulses with valid arrived.
It is a simple change that will enable the test to run with the real AST too. 'N' Can be a parameter as you suggested.

**Comment by tjaychen** [comment]
> 

sounds good @Jacob-Levy . do you want me to hold on the merge? Or do you prefer I merge it first, then you can run this test with your change?

**Comment by Jacob-Levy** [comment]
@tjaychen Go ahead and merge. 
I will code the change into a draft PR and let you adjust your test code into a new PR.

**Comment by tjaychen** [comment]
> @tjaychen Go ahead and merge. I will code the change into a draft PR and let you adjust your test code into a new PR.

sounds good!

**Comment by Jacob-Levy** [comment]
@tjaychen  [AST] USB Clock Calibrate Test Supports #14618 Added the support needed for the test

**Comment by tjaychen** [comment]
@Jacob-Levy you'll probably need to add some waivers for your recent changes.

https://reports.opentitan.org/prod/opentitan/github/presubmit/3285/20220826-150336/HEAD/reports/hw/top_earlgrey/lint/ascentlint/latest/report.html

**Comment by Jacob-Levy** [comment]
@tjaychen  NTIL does not have AscentLint tool and @msfschaffner was updating the waivers after my review.

**Comment by tjaychen** [comment]
yes that's right, but we have ascentlint blocking so that the submitters are able to fix issues themselves.
You can have a look at the warnings issued, and the waiver file [here](https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/ip/ast/lint/ast.waiver) to get a sense on how to add more.

**Pull request merged**
